// ssignal/lib/ssignal.es.js
class l extends EventTarget {
  #e;
  /**
   * Creates a new SSignal instance.
   * If the initial value is a Map or Set, it is wrapped in a reactive proxy that
   * dispatches change events on mutating calls.
   *
   * @param value - The initial value of the signal.
   */
  constructor(t) {
    super(), t instanceof Map || t instanceof Set ? this.#e = this.#t(t) : this.#e = t;
  }
  /**
   * Returns the current value of the signal.
   */
  get value() {
    return this.#e;
  }
  /**
   * Sets a new value for the signal. Accepts either a direct value or an updater
   * function that receives the previous value and returns the next one.
   * No event is dispatched when the new value is strictly equal to the current one.
   *
   * @param newValue - The next value, or a function `(prev: T) => T`.
   */
  set value(t) {
    const e = typeof t == "function" ? t(this.#e) : t;
    Object.is(e, this.#e) || (this.#e = e instanceof Map || e instanceof Set ? this.#t(e) : e, this.dispatchEvent(new CustomEvent("change", { detail: this.#e })));
  }
  /**
   * Registers a callback that is invoked whenever the signal value changes.
   * Returns an unsubscribe function that removes the listener when called.
   *
   * Optionally accepts an AbortSignal to cancel the subscription automatically.
   * If the signal is already aborted at call time, the callback is never registered.
   *
   * @param callback - Function called with the new value on each change.
   * @param options.signal - Optional AbortSignal to cancel the subscription.
   * @param options.immediate - If true, fires the callback synchronously with the current value before returning.
   * @returns A function that removes the subscription when called.
   *
   * @example
   * const controller = new AbortController();
   * signal.subscribe((v) => console.log(v), { signal: controller.signal });
   * controller.abort(); // unsubscribes
   *
   * @example
   * signal.subscribe((v) => render(v), { immediate: true }); // render called immediately with current value
   */
  subscribe(t, e) {
    if (e?.signal?.aborted) return () => {
    };
    const n = (i) => t(i.detail);
    this.addEventListener("change", n);
    const s = () => this.removeEventListener("change", n);
    return e?.signal && e.signal.addEventListener("abort", s, { once: !0 }), e?.immediate && t(this.#e), s;
  }
  /**
   * Registers a callback that is invoked only on the next signal value change.
   * Returns an unsubscribe function that can cancel the pending callback before
   * it fires.
   *
   * Optionally accepts an AbortSignal to cancel the one-time subscription
   * automatically. If the signal is already aborted at call time, the callback is
   * never registered.
   *
   * @param callback - Function called with the new value on the next change.
   * @param options.signal - Optional AbortSignal to cancel the subscription.
   * @returns A function that removes the pending one-time subscription.
   *
   * @example
   * const unsubscribe = signal.once((v) => console.log('first change:', v));
   * unsubscribe(); // cancels if the signal has not changed yet
   */
  once(t, e) {
    if (e?.signal?.aborted) return () => {
    };
    let n = !0;
    const s = () => {
      n && (n = !1, this.removeEventListener("change", i), e?.signal?.removeEventListener("abort", s));
    }, i = (r) => {
      s(), t(r.detail);
    };
    return this.addEventListener("change", i), e?.signal && e.signal.addEventListener("abort", s, { once: !0 }), s;
  }
  /**
   * Wraps a Map or Set in a Proxy that dispatches a change event after any mutating
   * operation, keeping read methods working transparently.
   */
  #t(t) {
    const e = t instanceof Map ? ["set", "delete", "clear"] : ["add", "delete", "clear"];
    return new Proxy(t, {
      get: (n, s) => {
        const i = Reflect.get(n, s);
        return e.includes(String(s)) ? (...r) => {
          const c = i.apply(n, r);
          return this.dispatchEvent(new CustomEvent("change", { detail: this.#e })), c;
        } : typeof i == "function" ? i.bind(n) : i;
      }
    });
  }
}
const h = Object.getOwnPropertyDescriptor(l.prototype, "value").set, u = new FinalizationRegistry((a) => a());
class o extends l {
  #e;
  /** @internal */
  constructor(t, e) {
    const n = () => t.map((r) => r.value);
    super(e(...n()));
    const s = new WeakRef(this), i = t.map(
      (r) => r.subscribe(() => {
        const c = s.deref();
        c && h.call(c, e(...n()));
      })
    );
    this.#e = () => i.forEach((r) => r()), u.register(this, this.#e, this);
  }
  get value() {
    return super.value;
  }
  /**
   * @throws {TypeError} Always — computed signals are read-only.
   */
  set value(t) {
    throw new TypeError("Cannot set the value of a computed signal. It is read-only.");
  }
  /**
   * Removes all subscriptions to source signals.
   * Call this when the computed signal is no longer needed to free memory.
   */
  dispose() {
    u.unregister(this), this.#e();
  }
}
function d(a, t) {
  return Array.isArray(a) ? new o(a, (...e) => t(e)) : new o([a], (e) => t(e));
}
export {
  o as ComputedSignal,
  d as computed,
  l as default
};
