// query-client/lib/query-item.es.js
var h = Object.defineProperty;
var y = (e, t, r) => t in e ? h(e, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : e[t] = r;
var s = (e, t, r) => y(e, typeof t != "symbol" ? t + "" : t, r);
const d = Symbol("query.item.metadata"), i = Symbol("query.item.data"), f = (e, t = "clone") => {
  if (e == null || typeof e != "object")
    return e;
  const r = (a) => {
    try {
      const n = globalThis.structuredClone;
      return typeof n == "function" ? n(a) : JSON.parse(JSON.stringify(a));
    } catch (n) {
      return a;
    }
  }, o = (a) => (Object.freeze(a), Object.getOwnPropertyNames(a).forEach((n) => {
    a[n] !== null && typeof a[n] == "object" && o(a[n]);
  }), a);
  switch (t) {
    case "clone":
      return r(e);
    case "freeze":
      return o(r(e));
    case "reference":
      return e;
    default:
      return e;
  }
};
var c, u;
u = d, c = i;
class p {
  constructor(t, { queryFn: r, staleTime: o = 0 }) {
    s(this, "queryFn");
    s(this, u);
    s(this, c);
    /**
     * Public accessor for the data which returns a cloned copy to prevent
     * accidental external mutation of the cached value.
     */
    s(this, "dataStrategy", "clone");
    this[i] = t, this.queryFn = r, this[d] = {
      dataCreatedAt: Date.now(),
      dataUpdatedAt: Date.now(),
      errorUpdatedAt: void 0,
      errorUpdateCount: 0,
      isInvalidated: !1,
      staleTime: o,
      timeoutId: void 0
    };
  }
  get data() {
    return f(this[i], this.dataStrategy);
  }
  setDataStrategy(t) {
    return this.dataStrategy = t, this;
  }
  /**
   * Time left in milliseconds before the item becomes stale. Never negative.
   */
  get timeLeftToStale() {
    const t = this[d].dataUpdatedAt + this[d].staleTime;
    return Math.max(0, t - Date.now());
  }
  getMetadata() {
    return this[d];
  }
  updateData(t) {
    return this[i] = t, this.getMetadata().dataUpdatedAt = Date.now(), this;
  }
  updateError() {
    return this.getMetadata().errorUpdatedAt = Date.now(), this.getMetadata().errorUpdateCount += 1, this;
  }
  invalidate() {
    return this[i] = void 0, this.getMetadata().isInvalidated = !0, this;
  }
  isStale() {
    return Date.now() - this.getMetadata().dataUpdatedAt > this.getMetadata().staleTime;
  }
}
export {
  p as QueryItem
};
